File: Algorithms on Graphs/week1_decomposition1/1_reachability/reachability.py
# Uses python3

import sys


def explore(i, adj, visited):
    visited[i] = 1

    for j in adj[i]:
        if visited[j] == 0:
            explore(j, adj, visited)


def reach(adj, x, y):
    # write your code here
    visited = [0] * (len(adj) + 1)

    explore(x, adj, visited)

    if visited[x] == 1 and visited[y] == 1:
        return 1

    return 0


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    # print(data)
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    # print(edges)
    x, y = data[2 * m:]
    adj = [[] for _ in range(n)]
    x, y = x - 1, y - 1
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)

    result = reach(adj, x, y)
    print(result)

File: Algorithms on Graphs/week1_decomposition1/1_reachability/Reachability.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class Reachability {
    private static int reach(ArrayList<Integer>[] adj, int x, int y) {
        //write your code here
        int visited[] = new int[adj.length];          // make array to keep track of visited nodes
        Arrays.fill(visited, 0);                      // initially set all values to 0

        explore(visited, adj, x);                     // start exploring from root node

        return visited[y];                            // return state of destination node if 1 then it's reachable
    }

    public static void explore(int[] visited, ArrayList<Integer>[] adj, int x)    // function to DFS graph
    {
        visited[x] = 1;                               // make node as visited

        for(int i : adj[x])                           // iterate through all adjacent nodes
        {
            if(visited[i] == 0)                       // if not visited
            {
                explore(visited, adj, i);             // visit recursively
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        System.out.println(reach(adj, x, y));
    }
}

